public/app/features/query/state/DashboardQueryRunner/UnifiedAlertStatesWorker.ts

```diff
@@ -0,0 +1,91 @@
+import { DashboardQueryRunnerOptions, DashboardQueryRunnerWorker, DashboardQueryRunnerWorkerResult } from './types';
+import { from, Observable } from 'rxjs';
+import { getBackendSrv } from '@grafana/runtime';
+import { catchError, map } from 'rxjs/operators';
+import { emptyResult, handleDashboardQueryRunnerWorkerError } from './utils';
+import { PromAlertingRuleState, PromRulesResponse } from 'app/types/unified-alerting-dto';
+import { AlertState, AlertStateInfo } from '@grafana/data';
+import { isAlertingRule } from 'app/features/alerting/unified/utils/rules';
+import { Annotation } from 'app/features/alerting/unified/utils/constants';
+
+export class UnifiedAlertStatesWorker implements DashboardQueryRunnerWorker {
+  canWork({ dashboard, range }: DashboardQueryRunnerOptions): boolean {
+    if (!dashboard.uid) {
+      return false;
+    }
+
+    if (range.raw.to !== 'now') {
+      return false;
+    }
+
+    return true;
+  }
+
+  work(options: DashboardQueryRunnerOptions): Observable<DashboardQueryRunnerWorkerResult> {
+    if (!this.canWork(options)) {
+      return emptyResult();
+    }
+
+    const { dashboard } = options;
+    return from(
+      getBackendSrv().get(
+        '/api/prometheus/grafana/api/v1/rules',
```

Comment:
feels like we need to pass dashboardId here and some for of SQL index to filter rules, fetching all rules for every dashboard is seems too costly, could be many thousands of rules and this is fetched every refresh

Reply:
Nice catch 👍

Reply:
Hmm, good point. I was hoping Grafana would to serve this result entirely from memory, but that does not seem to be the case.
There's also an alternative solution for this issue possible where we modify `/api/alerts/states-for-dashboard` endpoint in the backend to get states for ngalert alerts too instead of using the new alerting API here.
I'll consult with backend peeps

Reply:
Discussed with backend peeps, it will take a design doc and some time to produce an endpoint to return only rules for a particular dashboard. Next week I'll try to generate a few k rules and do some profiling to have a better idea of how much of a problem this is

Reply:
I generated 3000 grafana alert rules to see what happens.

API endpoint to get all rule statuses seems to be OK "on my machine", 111 KB / 460 ms., not really noticable.

![rules](https://user-images.githubusercontent.com/847684/132223772-2beb12c1-5870-444b-bc1c-8f1abde78851.png)

Backend struggles to evaluate this many rules though, seems to currently be beyond a practical limit. But that's not directly relevant to this PR.

 @torkelo  @hugohaggmark  do you think I could merge this and improve later, when an endpoint to retreive dashboard-specific rules is implemented, or would you rather wait?

Reply:
All green from me, @torkelo?

Reply:
not really, think we have a problem with this and AlertStatesWorker in that it executes for all dashboards on every refresh, even for dashboards with no alert rules. As these calls can be very expensive (even on play with minimal data/alert rules https://play.grafana.org/api/alerts/states-for-dashboard?dashboardId=14 takes 490ms